### DISCUSSION TITLE
Simple example with partial and final results for chunked audio stream?

### DISCUSSION THREAD

**Discussion by fquirin** [NVIDIA-NeMo/NeMo#4216] [Q&A] [answered]
Hi everybody,

NeMo is giving me really great results for audio file transcription :+1: (conformer ctc medium).
Now I'm trying to build a simple example for streaming audio with partial and final results, but I'm a bit lost :worried: .

I've seen the [speech_to_text_buffered_infer_ctc.py](https://github.com/NVIDIA/NeMo/blob/stable/examples/asr/asr_chunked_inference/ctc/speech_to_text_buffered_infer_ctc.py) example, the [Online_ASR_Microphone_Demo.ipynb](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Online_ASR_Microphone_Demo.ipynb) and the [transcribe_speech.py](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/transcribe_speech.py) but I'm having trouble isolating the relevant parts since there is a lot going on.

It seems there is no dedicated function like in other toolkits (Coqui or Vosk for example) but you have to wrap another class around the original 'ASRModel' (in my case 'EncDecCTCModel') and modify what is called 'the preprocessor'.

Maybe you can help me finish the following example code. Please note: I don't need any data evaluation, WER calculation, samplerate transformation, etc. etc., I just want to feed the audio and get a text result ^^:

```
import copy
import wave

import numpy as np
import torch
from omegaconf import OmegaConf
import nemo.collections.asr as nemo_asr
from nemo.utils import logging, model_utils

# Load the model 
model_file = "models/stt_en_conformer_ctc_medium.nemo"
model = nemo_asr.models.EncDecCTCModel.restore_from(restore_path=model_file, map_location="cpu")
model = model.to("cpu")    
# BTW: no matter how I tell the model to use CPU it always runs into the auto-detection warning later

# Modify preprocessor
cfg = copy.deepcopy(model._cfg)
OmegaConf.set_struct(cfg.preprocessor, False)  # enable config overwriting

# "some changes for streaming scenario" - no idea what this does
cfg.preprocessor.dither = 0.0
cfg.preprocessor.pad_to = 0

OmegaConf.set_struct(cfg.preprocessor, True)  # disable config overwriting
model.eval()

# TODO: Some magic using 'FrameBatchASR' or a completely new wrapper class ... ???
frame_asr = ???

# Load a wave file - 16khz mono PCM
wf = wave.open("audio/test.wav", "rb")

# Serve the wave file in chunks and do magic
chunk_size = 4000
while True:
    # Read some bytes
    frames = wf.readframes(chunk_size)

    # Process as long as we have data
    if len(frames) == 0:
        # TODO: finish and end
        final_result = ???
    else:
        # TODO: Convert to the right input (??) format
        data = np.frombuffer(frames, dtype=np.int16)

        # TODO: feed chunks get partial result
        partial_result = ???

```

I'd be happy about any additional info, hints or code examples :slightly_smiling_face: 
Ty,
Florian

**Comment by titu1994** [comment, marked as answer]
So as you've already guessed, Nemo ASR models are complex underneath the hood. I will spend some time next week to see a minimal script based on yours that works. But before that look at https://huggingface.co/spaces/smajumdar/nemo_conformer_rnnt_large_streaming

There's no special tricks applied here, it's the most inefficient inference method on chunks, and that works fine actually. These models in Nemo are not true streaming models, but offline models. We can make them work in streaming mode in multiple ways - buffered inference script above is one of those days. It's a more accurate form of the simple "predict full chunk, every chunk, and concat results" method I've used in the above demo. 

We do provide transcribe() method that takes in a file and output's text that does the entire pipeline inside it. The drawback to this is we don't support raw audio stream input. 

Since there's recently been request for it in both hf and other users, we will look into defining a transcribe_buffer() or something like that which accepts tensors and output's text, but it will take time. 

In the meantime, I will try a minimal script for transcribing a buffer based on your code and respond next week c

**Reply by titu1994** [reply]
Sidenote we will be getting true streaming Conformers soon*. @VahidooX is working on them at the moment.

**Reply by fquirin** [reply]
Thanks a lot for the info! :+1: 

> I will spend some time next week to see a minimal script based on yours that works

That would be awesome :slightly_smiling_face: . I was looking at [transcribe_partial_audio](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/utils/transcribe_utils.py) and it looks very useful, but its based on a data manifest or an audio files directory. Maybe I can find a way to rewrite that :thinking: 

> But before that look at https://huggingface.co/spaces/smajumdar/nemo_conformer_rnnt_large_streaming

The code looks like its simply using the default 'transcribe'. Are the "chunks" actually separate (temporary) recordings? Or am I missing something? :thinking: 

> Since there's recently been request for it in both hf and other users, we will look into defining a transcribe_buffer() or something like that which accepts tensors and output's text, but it will take time.

I think that will be very useful, yes! :slightly_smiling_face: 

> Sidenote we will be getting true streaming Conformers soon*. @VahidooX is working on them at the moment.

Great! :star:

**Reply by titu1994** [reply]
That code is basically a copy paste of what's inside model.transcribe(). Chunking is left upto outside engine, not done by any of our scripts.

**Reply by fquirin** [reply]
I see, ty. Since this would require some VAD based chunk splitting I think I can't apply it. My input is usually only one sentence and the partial results are more of a user-experience thing to get feedback as quickly as possible and start transcribing while the user is still speaking.

**Reply by titu1994** [reply]
We'd recommend Riva for that, but it won't work on a raspberry pi

**Reply by fquirin** [reply]
True, I mean from Nvidia's point of view I can totally under stand it but the GPU requirement is unfortunately a no-go for me. I'm very happy though that NeMo CTC medium models are pretty fast on Raspberry Pi :slightly_smiling_face:

**Reply by fquirin** [reply]
> In the meantime, I will try a minimal script for transcribing a buffer based on your code and respond next week c

Did you find some time to look into this already? :slightly_smiling_face:

**Reply by titu1994** [reply]
Nope, US holidays. We also have a release now so it might take a little while (2-3 weeks)

**Reply by fquirin** [reply]
ok np, ty for the info and enjoy your holidays :+1:

**Reply by fquirin** [reply]
Hope you've enjoyed holidays :slightly_smiling_face: , if you've time I'd still be interested in the example code :innocent:.
Also: anywhere to follow the "true streaming conformer" progress? :-)

**Reply by VahidooX** [reply]
Here is the PR for streaming Conformer: https://github.com/NVIDIA/NeMo/pull/3888
It has a small issue to get fixed and going to get merged next week.

**Reply by fquirin** [reply]
Awesome! I'll definitely have a closer look soon 😎

**Reply by fquirin** [reply]
Hi @VahidooX , I just read the new docs section about the [streaming conformer model](https://github.com/NVIDIA/NeMo/blob/main/docs/source/asr/models.rst#cache-aware-streaming-conformer) and I was wondering how the `chunk-aware look-ahead` mode works. The descriptions says "self-attention layers would be able to see all the tokens in their corresponding chunk", but what if a word for example is split between one or multiple chunks? Do new chunks still improve the "old" result. To be honest this is something I always asked myself when trying to understand streaming ASR, how do you handle overlap without constantly recalculating old data? 🤔

**Comment by BakingBrains** [comment]
Any updated code on getting partial and final result @fquirin.

Thanks

**Reply by fquirin** [reply]
I haven't checked the lib for a while, but would be interested in the answer as well ^^.
Maybe @titu1994 or @VahidooX can update us on the state of partial results and streaming models?
